### PR TITLE
Enroll and pay later: Load dashboard page without reloading

### PR DESCRIPTION
**Pull request by amir-qayyum-khan**
#### What are the relevant tickets?
fixes https://github.com/mitodl/micromasters/issues/1445

#### What's this PR do?
- Show inline success message when user press enroll and pay later button,
- It also loads dashboard page without reloading whole page

#### How should this be manually tested?
- Open `http://192.168.99.100:8079/dashboard/` and press enroll and pay later.

#### Screenshots (if appropriate)
screen shot 2017-03-13 at 3 57 59 pm

### CODE REVIEW

**Comment by amir-qayyum-khan** [comment]
@noisecapella can you verify that i am on right track?

File: static/js/components/dashboard/CourseListCard_test.js [hunk 1]
```diff
@@ -58,6 +59,7 @@ describe('CourseListCard', () => {
             coursePrice={coursePrice}
             addCourseEnrollment={() => Promise.resolve()}
             prices={prices}
+            ui={{ showEnrollPayLaterSuccess: false }}
```
Comment (amir-qayyum-khan) [inline static/js/components/dashboard/CourseListCard_test.js]: @noisecapella i am getting flow warnings here, it is excepting all from `UIState`;

Comment (amir-qayyum-khan) [inline static/js/components/dashboard/CourseListCard_test.js]: fixed

File: static/js/reducers/dashboard.js [hunk 2]
```diff
@@ -36,7 +36,11 @@ export const dashboard = (state: DashboardsState = INITIAL_DASHBOARDS_STATE, act
   case REQUEST_DASHBOARD: // eslint-disable-line no-case-declarations
     let newBaseState = R.dissoc(username, state);
     if (action.payload === true) {
-      return updateDashboardState(newBaseState, username, INITIAL_DASHBOARD_STATE);
+      return updateDashboardState(
+        newBaseState,
+        username,
+        _.merge({}, INITIAL_DASHBOARD_STATE, { fetchStatus: null })
```
Comment (aliceriot) [inline static/js/reducers/dashboard.js]: what's the reason for this change? the `newBaseState` shouldn't have anything saved in it under the `username` and `INITIAL_DASHBOARD_STATE` doesn't have a `fetchStatus` key on it.

File: static/js/actions/course_enrollments.js [hunk 3]
```diff
@@ -22,9 +23,10 @@ export const addCourseEnrollment = (courseId: string): Dispatcher<*> => {
     dispatch(requestAddCourseEnrollment(courseId));
     return api.addCourseEnrollment(courseId).
       then(() => {
-        dispatch(receiveAddCourseEnrollmentSuccess());
-        dispatch(fetchDashboard(SETTINGS.user.username));
-        dispatch(fetchCoursePrices());
+        dispatch(receiveAddCourseEnrollmentSuccess(courseId));
+        dispatch(fetchDashboard(SETTINGS.user.username, true));
+        dispatch(fetchCoursePrices(true));
+        dispatch(showEnrollPayLaterSuccessMessage());
```
Comment (aliceriot) [inline static/js/actions/course_enrollments.js]: probably you want the `.then` callback here to return this call (`dispatch(showEnrollPayLaterSuccessMessage());`)

File: static/js/actions/course_enrollments.js [hunk 4]
```diff
@@ -33,3 +35,16 @@ export const addCourseEnrollment = (courseId: string): Dispatcher<*> => {
...
+    dispatch(showEnrollPayLaterSuccess(true));
+    let promise = new Promise(function(resolve) {
+      window.setTimeout(function() {
+        dispatch(showEnrollPayLaterSuccess(false));
+        resolve('done!');
+      }, 9000);
+    });
+    return promise;
+  };
+};
```
Comment (aliceriot) [inline static/js/actions/course_enrollments.js]: this function would probably be better located in an actions file

**Comment by aliceriot** [comment]
@amir-qayyum-khan I left a few comments for some small things I noticed. I think @noisecapella is much more familiar with the code at hand, though, so probably is a better person for that!

**Comment by noisecapella** [comment]
@amir-qayyum-khan It looks like you are on the right track though I haven't done a functionality test. Can you add unit tests?

**Comment by amir-qayyum-khan** [comment]
ok @noisecapella

**Comment by amir-qayyum-khan** [comment]
Done @noisecapella, please review.

File: static/js/reducers/ui_test.js [hunk 5]
```diff
@@ -218,6 +219,14 @@ describe('ui reducers', () => {
     });
   });
 
+  describe('show enroll pay later success alert', () => {
+    it('should let you set the pay later success alert', () => {
+      assertReducerResultState(
+        showEnrollPayLaterSuccess, ui => ui.showEnrollPayLaterSuccess, {}
```
Comment (amir-qayyum-khan) [inline static/js/reducers/ui_test.js]: @noisecapella this is failing on CI, but it is working fine on my machine.
```AssertionError: expected {} to deeply equal { course_id: 'course-v1:48' }
      + expected - actual
      -{}
      +{
      +  "course_id": "course-v1:48"
      +}
      ```

File: static/js/actions/course_enrollments.js (hunk 4 above)
Comment (amir-qayyum-khan) [inline static/js/actions/course_enrollments.js]: it is in actions actions/course_enrollments.js,

by the way done with addressing ur comments

**Comment by noisecapella** [comment]
It looks like there's actually two dashboard refreshes happening, when the user clicks the enroll button:

screenshot from 2017-03-15 15-32-51

This is why the spinner is still showing up even though `noSpinner` is true for one of the actions

File: static/js/actions/course_enrollments.js [hunk 6]
```diff
@@ -33,3 +36,16 @@ export const addCourseEnrollment = (courseId: string): Dispatcher<*> => {
       });
   };
 };
+
+export const showEnrollPayLaterSuccessMessage = (courseId: string): Dispatcher<*> => {
+  return (dispatch: Dispatch) => {
+    dispatch(showEnrollPayLaterSuccess({course_id: courseId}));
```
Comment (noisecapella) [inline static/js/actions/course_enrollments.js]: If the only thing we care about is the course key, can we just pass `courseId` here by itself instead of within an object?

File: static/js/actions/course_enrollments.js [hunk 7]
```diff
@@ -33,3 +36,16 @@ export const addCourseEnrollment = (courseId: string): Dispatcher<*> => {
...
+    let promise = new Promise(function(resolve) {
+      window.setTimeout(function() {
+        dispatch(showEnrollPayLaterSuccess(INITIAL_ENROLL_PAY_LATER_STATE));
```
Comment (noisecapella) [inline static/js/actions/course_enrollments.js]: Can you add a test for this code?

File: static/js/reducers/index.js [hunk 8]
```diff
@@ -211,10 +211,14 @@ const INITIAL_COURSE_PRICES_STATE: CoursePricesState = {
 export const prices = (state: CoursePricesState = INITIAL_COURSE_PRICES_STATE, action: Action<any, null>) => {
   switch (action.type) {
   case REQUEST_COURSE_PRICES:
-    return {
-      ...state,
-      fetchStatus: FETCH_PROCESSING
-    };
+    if (action.payload === true) {
+      return INITIAL_COURSE_PRICES_STATE;
```
Comment (noisecapella) [inline static/js/reducers/index.js]: Can you add a test for this?

**Comment by amir-qayyum-khan** [comment]
@noisecapella added tests, i need your help in investigation why spinner is still running.
can you give me headsup

**Comment by noisecapella** [comment]
I investigated and it looks like this is happening because the `noSpinner` option clears `fetchStatus`, then `DashboardPage` interprets that to mean that the dashboard page fetch has not happened yet. I think the best way to fix this is to change the reducer to add `noSpinner` explicitly to the state if it's `true` and leave `fetchStatus` alone. Then `DashboardPage` will need to look at both `fetchStatus` and `noSpinner` in order to determine whether to show the spinner. Does that make sense?

**Comment by amir-qayyum-khan** [comment]
Fixed @noisecapella

File: static/js/reducers/dashboard.js [hunk 9]
```diff
@@ -34,7 +35,11 @@ export const dashboard = (state: DashboardsState = INITIAL_DASHBOARDS_STATE, act
   case REQUEST_DASHBOARD: // eslint-disable-line no-case-declarations
     let newBaseState = R.dissoc(username, state);
     if (action.payload === true) {
-      return updateStateByUsername(newBaseState, username, INITIAL_DASHBOARD_STATE);
+      return updateStateByUsername(
+        newBaseState,
+        username,
+        _.merge({}, INITIAL_DASHBOARD_STATE, { noSpinner: true })
```
Comment (noisecapella) [inline static/js/reducers/dashboard.js]: We need to also set `fetchStatus` here. Instead of having the `if (action.payload === true)` we could just have

    {
      fetchStatus: FETCH_PROCESSING,
      noSpinner: action.payload,
    }

**Comment by amir-qayyum-khan** [comment]
Done @noisecapella

File: static/js/reducers/dashboard.js (hunk 9 above)
Comment (amir-qayyum-khan) [inline static/js/reducers/dashboard.js]: if i do this spinner start spinning

File: static/js/reducers/dashboard.js [hunk 10]
```diff
@@ -34,7 +35,13 @@ export const dashboard = (state: DashboardsState = INITIAL_DASHBOARDS_STATE, act
   case REQUEST_DASHBOARD: // eslint-disable-line no-case-declarations
     let newBaseState = R.dissoc(username, state);
     if (action.payload === true) {
-      return updateStateByUsername(newBaseState, username, INITIAL_DASHBOARD_STATE);
+      return updateStateByUsername(
+        newBaseState,
+        username,
+        _.merge({}, INITIAL_DASHBOARD_STATE, {
+          noSpinner: true
```
Comment (noisecapella) [inline static/js/reducers/dashboard.js]: We should have `fetchStatus: FETCH_PROCESSING` here too, and because the two branches of this if statement are similar, you can just set `noSpinner: action.payload`.

Regarding your other comment on the dashboard spinner still spinning, we should change that code to also check `noSpinner`

File: static/js/containers/DashboardPage.js [hunk 11]
```diff
@@ -262,14 +262,14 @@ class DashboardPage extends React.Component {
 
   fetchDashboard() {
     const { dashboard, dispatch } = this.props;
-    if (dashboard.fetchStatus === undefined) {
+    if (dashboard.fetchStatus === undefined && dashboard.noSpinner === false) {
```
Comment (noisecapella) [inline static/js/containers/DashboardPage.js]: If you set `fetchStatus` in all cases as described in my previous comment, you shouldn't need to check `noSpinner` at all here or below.

File: static/js/reducers/course_prices.js [hunk 12]
```diff
@@ -21,22 +21,36 @@ import type {
...
-      R.dissoc(username, state),
-      username,
-      _.merge({}, INITIAL_COURSE_PRICES_STATE, { fetchStatus: FETCH_PROCESSING })
-    );
+    if (action.payload === true) {
+      return updateStateByUsername(
+        R.dissoc(username, state),
+        username,
+        _.merge({}, INITIAL_COURSE_PRICES_STATE, {
+          noSpinner: true
```
Comment (noisecapella) [inline static/js/reducers/course_prices.js]: Please set `fetchStatus` here too and merge the two if branches like I suggested in the dashboard reducer

File: static/js/containers/DashboardPage.js [hunk 13]
```diff
@@ -694,6 +694,7 @@ class DashboardPage extends React.Component {
               coursePrice={coursePrice}
               prices={calculatedPrices}
               key={program.id}
+              ui={ui}
               openFinancialAidCalculator={this.openFinancialAidCalculator}
               addCourseEnrollment={this.addCourseEnrollment}
               openCourseContactDialog={this.openCourseContactDialog}
```
Comment (noisecapella) [inline static/js/containers/DashboardPage.js]: Look below in `render()`. You should be checking `noSpinner` there in addition to `fetchStatus`

**Comment by amir-qayyum-khan** [comment]
done @noisecapella

File: static/js/reducers/course_prices.js [hunk 14]
```diff
@@ -21,22 +21,37 @@ import type {
...
 export const prices = (state: CoursePriceReducerState = {}, action: Action<any, string>) => {
   const { meta: username } = action;
   switch (action.type) {
   case REQUEST_COURSE_PRICES:
-    return updateStateByUsername(
-      R.dissoc(username, state),
-      username,
-      _.merge({}, INITIAL_COURSE_PRICES_STATE, { fetchStatus: FETCH_PROCESSING })
-    );
+    if (action.payload === true) {
```
Comment (noisecapella) [inline static/js/reducers/course_prices.js]: Please remove this if statement. Since the only difference between the two cases is the `noSpinner`, you can have this instead:

    return updateStateByUsername(
      R.dissoc(username, state),
      username,
      _.merge({}, INITIAL_COURSE_PRICES_STATE, {
         noSpinner: action.payload,
        fetchStatus: FETCH_PROCESSING
      })
    );

Please do the same for the dashboard reducer

Comment (amir-qayyum-khan) [inline static/js/reducers/course_prices.js]: done @noisecapella

**Comment by noisecapella** [comment]
@amir-qayyum-khan The spinner is gone but it seems like the dashboard page temporarily disappears while the loading happens. I think we need to update the reducer to preserve the old data in the case that `noSpinner` is true. Can you make the change to the prices and dashboard reducers?

**Comment by amir-qayyum-khan** [comment]
fixed @noisecapella , please take a look
